### PR TITLE
3.6 - Add a deprecation warning helper.

### PR DESCRIPTION
**Pull request by markstory**
This will make adding framework deprecation warnings we plan to add in 3.6.x simpler as they will be easier to write.

### CODE REVIEW

**Comment by chinpei215** [comment]
I think the function would report unhelpful line number and file name.

**Comment by inoas** [comment]
```php
<?php

    function deprecationWarning($message)
    {
        trigger_error($message, E_USER_DEPRECATED);
    }

deprecationWarning('here');
```

PHP Deprecated:  here in /Users/leo/Downloads/test-error.php on line 5

Deprecated: here in /Users/leo/Downloads/test-error.php on line 5

**Comment by markstory** [comment]
@chinpei215 The HTML error output will include a backtrace. Would you want to see the backtrace in the deprecation warning message as well?

**Comment by dereuromark** [comment]
the log entries would be useful to have the right location and line number.

**Comment by markstory** [comment]
@dereuromark Logging is something that the error handler already does if its configured correctly.

**Comment by chinpei215** [comment]
@markstory But people would have to click links to show stacktraces when they encountered errors.
deprecations

**Comment by markstory** [comment]
@chinpei215 Fair, but they would have to do that anyways as the raw `trigger_error` calls would point to code inside cakephp not their application. We could do what @dereuromark suggested and 'jump up' two stack frames to hopefully get the right line.

**Comment by chinpei215** [comment]
@markstory ah sorry you are right. and it might be nice if we could do that.

**Comment by markstory** [comment]
I think generating a backtrace 2 levels up is reasonable. However, I would want to guard that kind of logic with a debug check. Does that seem reasonable?

**Comment by chinpei215** [comment]
Sounds reasonable to me.

**Comment by markstory** [comment]
Error messages now have file/lines displayed.

**Comment by gmponos** [comment]
IMHO I see more trouble caused rather than value added.

**Comment by markstory** [comment]
@gmponos If we use trigger_error() without a wrapper, the errors will not be able to indicate what application code is using the deprecated functions.

**Comment by gmponos** [comment]
> If we use trigger_error() without a wrapper, the errors will not be able to indicate what application code is using the deprecated functions.

If that's the goal then fine. I maybe rushed judging the goal of this PR because of the initial commit where the only thing I saw was a "proxy" function...

But still your comment raises some more questions for me and I am sorry for that... maybe because I don't follow all discussions.

I believe you are using an ErrorHandler on a CakePHP application isn't this possible to be handled there? Hope I didn't misunderstood and sorry for any trouble.

**Comment by markstory** [comment]
> I believe you are using an ErrorHandler/Debugger on a CakePHP application isn't this possible to be handled there?

We are using an error handler, but that error handler relies on errors coming out of `trigger_error()`.  While the current ErrorHandler does show a stack trace in the HTML output, the CLI output does not include stack trace information. Having to click into the stack trace for HTML messages, and having not hints at all in CLI contexts are some of the reasons I think this function could be useful to people upgrading.

File: src/Core/functions.php [hunk 1]
```diff
@@ -254,10 +254,28 @@ function env($key, $default = null)
...
      *
      * @param string $message The message to output as a deprecation warning.
+     * @param int $stackFrame The stack frame to include in the error. Defaults to 2
+     *   as that should point to application/plugin code.
      * @return void
      */
-    function deprecationWarning($message)
+    function deprecationWarning($message, $stackFrame = 2)
     {
+        if (!Configure::read('debug')) {
```
Comment (dereuromark) [inline src/Core/functions.php]: If we do this, then we wont be able to find any left overs in production.

Comment (markstory) [reply src/Core/functions.php]: Yes, but I'm not comfortable adding potentially 100s of calls to `debug_backtrace()` to people's applications in production mode. Would it be better to check the error level here instead?

Comment (dereuromark) [reply src/Core/functions.php]: Yeah, I think so.

Comment (inoas) [reply src/Core/functions.php]: I think we should have a separate flag for this. Something along `log3xDeprecations` (default true if not set), which should go into a dedicated log file. That way it is easy to spot in production and easy to disable if you just want to run your app and not update it.

File: src/Core/functions.php [hunk 2]
```diff
@@ -248,3 +248,34 @@ function env($key, $default = null)
...
+     * @param int $stackFrame The stack frame to include in the error. Defaults to 2
+     *   as that should point to application/plugin code.
+     * @return void
+     */
+    function deprecationWarning($message, $stackFrame = 2)
+    {
+        if (!(error_reporting() & E_USER_DEPRECATED)) {
+            return;
+        }
+        $trace = debug_backtrace();
```
Comment (ravage84) [inline src/Core/functions.php]: May be a white line above

File: tests/TestCase/Core/FunctionsTest.php [hunk 3]
```diff
@@ -42,4 +42,39 @@ public function testEnv()
         $this->assertEquals('0', env('ZERO'));
         $this->assertEquals('0', env('ZERO', '1'));
     }
+
+    /**
+     * Test error messages coming out when debug is on
```
Comment (ravage84) [inline tests/TestCase/Core/FunctionsTest.php]: "Test error messages coming out when debug is on, manually setting the stack frame"

File: tests/TestCase/Core/FunctionsTest.php [hunk 4]
```diff
@@ -42,4 +42,39 @@ public function testEnv()
...
+     * @expectedExceptionMessage This is going away - [internal], line: ??
+     */
+    public function testDeprecationWarningEnabled()
+    {
+        error_reporting(E_ALL);
+        deprecationWarning('This is going away', 1);
+    }
+
+    /**
+     * Test error messages coming out when debug is on
```
Comment (ravage84) [inline tests/TestCase/Core/FunctionsTest.php]: "Test error messages coming out when debug is on, not setting the stack frame manually"

**Review by ravage84 — COMMENTED** [review]
Beautiful!

**Pull request merged**